### PR TITLE
Dsp

### PR DESCRIPTION
**Pull request by iliaplatone**

### CODE REVIEW

**Comment by iliaplatone** [comment]
@knro I can't upload the matrix data, a BLOB for the convolution plugin. Can you help me please?
The file is libs/indibase/dsp/convolution.cpp
I put a draft for a fits parser into ISNewBLOB there.

**Comment by knro** [comment]
@iliaplatone matrix data is in that file? What error do you git when uploading?

**Comment by iliaplatone** [comment]
ISNewBLOB simply gets not called

**Comment by iliaplatone** [comment]
Also wavelets values are not updated

**Comment by knro** [comment]
Ok, can you tell me what steps do I need to replicate/test this issue with ISNewBlob and the wavelet values?

**Comment by iliaplatone** [comment]
BLOB upload:
You have to start CCD Simulator and enable, under signal processing tab, convolution.
This enables a RW BLOB, a small image to convolute with the whole image received from camera.
Since file parsing is not tested neither complete I couldn't test about crashes etc..
Wavelets:
When enabled wavelets in Signal Processing I can't change values. This also uses convolution from libdsp so It will be slow maybe

File: libs/indibase/indiccd.cpp [hunk 1]
```diff
@@ -1895,6 +1928,12 @@ bool CCD::ExposureComplete(CCDChip * targetChip)
     // Reset POLLMS to default value
     POLLMS = getPollingPeriod();
 
+    if(HasDSP()) {
+        uint8_t* buf = (uint8_t*)malloc(PrimaryCCD.getFrameBufferSize());
```
Comment (knro) [inline libs/indibase/indiccd.cpp]: Use static_cast<uint8_t*>(...)

File: libs/indibase/dsp/manager.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
+﻿/*******************************************************************************
+  Copyright(c) 2017 Jasem Mutlaq. All rights reserved.
+
+ Connection Plugin Interface
```
Comment (knro) [inline libs/indibase/dsp/manager.h]: Correct header.

File: libs/indibase/dsp/manager.cpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    convolution = new Convolution(dev);
+    transforms = new Transforms(dev);
+    spectrum = new Spectrum(dev);
+    histogram = new Histogram(dev);
+    wavelets = new Wavelets(dev);
+}
+
+Manager::~Manager()
+{
+    convolution->~Convolution();
```
Comment (knro) [inline libs/indibase/dsp/manager.cpp]: Why is calling the destructor explicitly required?

File: libs/indibase/dsp/dspinterface.h [hunk 4]
```diff
@@ -78,29 +85,24 @@ class Interface
...
     private:
-                double primaryFocalLength, primaryAperture;
-        double RA, Dec;
-        double Lat, Lon, El;
-        double MPSAS;
-        char processedFileName[MAXINDINAME];
-        char processedFileExtension[MAXINDIFORMAT];
-        int BufferSizesQty;
-        int *BufferSizes;
+        long BufferSizesQty;
```
Comment (knro) [inline libs/indibase/dsp/dspinterface.h]: If you can use uint32_t instead of long, that would be better.

**Comment by knro** [comment]
ok is there an example of what kind of small image I can use? and what output to expect?

**Comment by knro** [comment]
So in ISNewBLOB(..) in ccd_simulator.cpp, you simply need to change the function to say this:

ccdSim->ISNewBLOB(...) this way it gets passed and then propagated upwards

**Comment by knro** [comment]
Can you please also add the spectrograph simulator to drivers.xml?

**Comment by iliaplatone** [comment]
Try with  a small FIT (10x10 pixels for example) with a B/W star shape. Each star should output with this shape aberration.

**Pull request merged**